src/iterators/inverted_index_iterator.c

```diff
@@ -60,6 +61,118 @@ size_t InvIndIterator_NumEstimated(QueryIterator *base) {
   return it->idx->numDocs;
 }
 
+static ValidateStatus EmptyCheckAbort(QueryIterator *base) {
+  return VALIDATE_OK;
+}
+
+static ValidateStatus NumericCheckAbort(QueryIterator *base) {
+  NumericInvIndIterator *nit = (NumericInvIndIterator *)base;
+  InvIndIterator *it = (InvIndIterator *)base;
+
+  if (!it->sctx) {
+    return VALIDATE_OK;
+  }
+
+  // For numeric fields, we need to get the field name from the filter context
+  // We need to use the field spec directly from the numeric filter
+  const NumericFilter *filter = it->decoderCtx.filter;
+  if (!filter) {
+    // No filter means we can't check for revision changes
+    return VALIDATE_OK;
+  }
+
+  RedisModuleString *numField = IndexSpec_GetFormattedKey(it->sctx->spec, filter->fieldSpec, INDEXFLD_T_NUMERIC);
+  NumericRangeTree *rt = openNumericKeysDict(it->sctx->spec, numField, DONT_CREATE_INDEX);
+
+  if (!rt || rt->revisionId != nit->revisionId) {
+    // The numeric tree was either completely deleted or a node was split or removed.
+    // The cursor is invalidated.
+    base->isAborted = true;
+    return VALIDATE_ABORTED;
+  }
+
+  return VALIDATE_OK;
+}
+
+static ValidateStatus TermCheckAbort(QueryIterator *base) {
+  InvIndIterator *it = (InvIndIterator *)base;
+  if (!it->sctx) {
+    return VALIDATE_OK;
+  }
+  InvertedIndex *idx = Redis_OpenInvertedIndex(it->sctx, base->current->data.term.term->str,
+      base->current->data.term.term->len, false, NULL);
+  if (!idx || idx == TRIEMAP_NOTFOUND || it->idx != idx) {
+    // The inverted index was collected entirely by GC.
+    // All the documents that were inside were deleted and new ones were added.
+    // We will not continue reading those new results and instead abort reading
+    // for this specific inverted index.
+    base->isAborted = true;
+    return VALIDATE_ABORTED;
+  }
+  return VALIDATE_OK;
+}
+
+static ValidateStatus TagCheckAbort(QueryIterator *base) {
+  TagInvIndIterator *it = (TagInvIndIterator *)base;
+  if (!it->base.sctx) {
+    return VALIDATE_OK;
+  }
+  size_t sz;
+  InvertedIndex *idx = TagIndex_OpenIndex(it->tagIdx, base->current->data.term.term->str,
+      base->current->data.term.term->len, false, &sz);
+  if (!idx || idx == TRIEMAP_NOTFOUND || it->base.idx != idx) {
```

Comment:
Cannot return `NULL`
```suggestion
  if (idx == TRIEMAP_NOTFOUND || it->base.idx != idx) {
```